[Step2]경로 조회 기능 PR 요청 드립니다.

re-mora opened the pull request:
안녕하세요 :) 교육생 최용석입니다 😄

2단계 `경로 조회 기능` 미션 구현이 완료되어, 리뷰 요청드립니다.

미션에서 주어진 대로 `PathFinder`라는 도메인 객체를 중심으로 최단 경로 조회 기능을 구현하였는데,
PathFinder는 별도의 상태를 가지지 않아도 될 것 같아 Utils성의 static 메소드로 구현해 보았습니다.
또한, PathFinder, PathService, PathAcceptanceTest에 각각 단위테스트와 인수테스트를 작성해 보았습니다.

이번 미션 역시 미흡한 부분이 있다면 피드백 부탁드립니다 :)

re-mora committed:
docs: step2 요구사항 정리 및 TODO List 작성

re-mora committed:
test: jgrapht에 대한 학습 테스트 작성

re-mora committed:
test: 최단 경로 조회를 위한 인수테스트 작성

- 지하철 최단 경로 조회
- 출발역과 도착역이 같은 경우
- 출발역과 도착역이 연결되어 있지 않은 경우
- 출발역이나 도착역이 존재하지 않는 경우

re-mora committed:
chore(polish): 테스트에 필요한 파라미터 설정 부 위치를 하단으로 이동

re-mora committed:
feat: 최단 경로 조회 기능을 제공하는 도메인 구현

기능 구현
- 주어진 노선에 포함된 역과 구간을 이용하여 그래프 구성
- 구성된 그래프 내에서 출발지로부터 도착지까지의 최단 거리 경로 조회

예외 처리
- 출발지와 도착지가 같은 경우 에외 처리
- 출발지 혹은 도착지가 전체 노선에 포함된 역이 아닌 경우 예외 처리
- 출발지와 도착지 사이에 간선이 없는 경우 예외 처리

re-mora committed:
fix: 깨지는 jgraph 관련 테스트 수정

re-mora committed:
fix: rebase 과정에서 발생한 충돌 수정

re-mora committed:
fix: rebase 과정에서 발생한 충돌 수정

re-mora committed:
test: PathServiceTest에 대한 TC 작성

re-mora committed:
Merge branch 'mission/step2-preview' into mission/step2

dhmin5693 commented on src/main/java/nextstep/subway/line/dto/ShortestPathResponse.java:
```diff
@@ -0,0 +1,31 @@
+
+public class ShortestPathResponse {
+    private int distance;
+    private List<Station> stations;
```
외부로 노출되는 DTO 객체에 엔티티가 노출되었습니다.

JPA entity는 자바 코드이지만 데이터베이스의 테이블 하나를 구현하였다는 특이점을 지니고 있습니다.
데이터베이스의 구성 요소는 중요한 비밀 중 하나이죠. 이렇게 외부로 노출하시면 안됩니다.
게다가 데이터베이스는 한 번 배포되면 컬럼을 추가하거나 제거하는게 쉽지 않습니다.
이 API에 특정한 데이터를 더 추가해달라는 요구사항이 생겼다고 가정해보겠습니다.
그렇다면 `Station` 테이블에 컬럼을 추가하실 건가요?
그러기 쉽지 않다는걸 이미 잘 알고 계실겁니다.

레이어드 아키텍처 구조의 논리로 따져보자면 엔티티는 presentation 계층에 위치할 수 없습니다.
레이어드 아키텍처에서는 상위 계층이 하위 계층을 참조할 수 있으나 그 역은 불가능하도록 제한하고 있습니다.
presentation 계층은 가장 상위 계층이며 controller와 외부 통신용 DTO 등이 포함됩니다.
엔티티는 최하위 계층인 persistence 계층에 위치하고 있습니다.
따라서 `Station`이 dto에 들어가면 계층을 침범한 것이 됩니다.
컨트롤러가 서비스를 참조하지만 서비스는 컨트롤러를 참조하는 방식으로 코드를 작성하지 않는다는걸 생각해보시면 이해하기 쉽습니다.

---

미션에서 요구하는 stations의 스펙에는 createdAt이 포함되어 있습니다.

dhmin5693 commented on src/main/java/nextstep/subway/path/ui/PathController.java:
```diff
@@ -0,0 +1,34 @@
+    }
+
+    @GetMapping
+    public ResponseEntity findShortestPath(
```
```suggestion
    public ResponseEntity<ShortestPathResponse> findShortestPath(
```

가급적이면 제네릭 형식을 맞춰주세요.
확인도 편하고 코드 정합성에도 도움이 됩니다.

dhmin5693 commented on src/main/java/nextstep/subway/path/domain/PathFinder.java:
```diff
@@ -0,0 +1,87 @@
+    }
+
+    private static void addVertexByStation(List<Station> allLinesStations) {
+        allLinesStations.forEach(it -> graph.addVertex(it));
```
여기서 매번 static 변수에 값을 포함해주고 있어서 이 클래스를 static 용으로는 사용하지 않는게 좋겠네요.
한 번에 많은 스레드가 접근한다면 동시성 문제가 발생할 가능성이 높습니다.
게다가 `Line` 이 변경되거나 제거되는 케이스도 대처가 안됩니다.

dhmin5693 commented on src/main/java/nextstep/subway/path/domain/PathFinder.java:
```diff
@@ -0,0 +1,87 @@
+public class PathFinder {
+    public static final String SOURCE_AND_TARGET_IS_EQUAL_ERROR = "출발지와 도착지는 같을 수 없습니다.";
+    public static final String SOURCE_OR_TARGET_IS_NOT_CONTAINS_ALL_LINE_STATION_ERROR = "대상 노선에서 해당역을 찾을 수 없습니다.";
+    static WeightedMultigraph<Station, DefaultWeightedEdge> graph;
+    static {
+        graph = new WeightedMultigraph(DefaultWeightedEdge.class);
+    }
+
+    public static ShortestPathResponse findShortestPath(List<Line> allLines, Station source, Station target) {
+        validate(allLines, source, target);
+        addSectionsToGraph(allLines);
+
+        DijkstraShortestPath dijkstraShortestPath = new DijkstraShortestPath(graph);
+        GraphPath path = dijkstraShortestPath.getPath(source, target);
+        checkResultIsNull(path);
+        return ShortestPathResponse.from(path);
```
`WeightedMultigraph`, `DijkstraShortestPath`, `GraphPath` 등 제네릭 타입을 지원하는 클래스는 꼭 제네릭으로 명시해서 정합성을 지켜주세요.

dhmin5693 commented on src/main/java/nextstep/subway/path/ui/PathController.java:
```diff
@@ -0,0 +1,34 @@
+
+    @ExceptionHandler(IllegalArgumentException.class)
+    public ResponseEntity handleIllegalArgsException(IllegalArgumentException e) {
+        return ResponseEntity.badRequest().build();
```
오류 메시지 정도는 담아서 줘도 괜찮지 않을까요?
이렇게만 던지면 디버깅이 어려울 수 있습니다.

dhmin5693 commented on src/test/java/nextstep/subway/path/application/PathServiceTest.java:
```diff
@@ -0,0 +1,184 @@
+        given(stationService.findStationById(역번호)).willReturn(반환역);
+    }
+
+    private void 전체_노선_조회_동작_정의() {
+        given(lineRepository.findAll()).willReturn(전체_노선);
+    }
```
이 메소드처럼 자주 정의하는 형태는 mock 보다는 fake 를 활용해보시는 방법도 좋습니다.
https://tecoble.techcourse.co.kr/post/2020-09-19-what-is-test-double/

dhmin5693 commented on src/main/java/nextstep/subway/line/dto/ShortestPathResponse.java:
```diff
@@ -0,0 +1,31 @@
+package nextstep.subway.line.dto;
```
ine보다는 path 패키지 이하가 더 어울리지 않을까요?

dhmin5693 reviewed (changes requested):
미션이 너무 빠르다 싶었는데 미리 작성한 뒤 리베이스를 해주셨군요 😅
배움을 향한 용식님의 열정, 너무 좋네요.

몇 가지 피드백을 남겨두었으니 확인 후 재요청 부탁드립니다. 🙏

re-mora committed:
docs: 코드 리뷰 피드백 내용 정리

re-mora committed:
docs: 엔티티를 응답 객체로 반환하는 코드 수정

최단 경로 조회 응답으로 `Station` 엔티티를 반환함으로써 발생할 수 있는 문제
- 데이터베이스의 테이블과 엔티티를 반환하는 경우 테이블 명세가 외부로 노출
- 응답 항목이 추가/변경 되는 요구사항에 테이블과 매핑된 엔티티로는 유연하게 대처할 수 없음
- 해당 엔티티가 양방향 연관관계를 가지고, toString이 구현되어 있는 경우 순환 참조로 인해 `StackOverFlow ` 발생할 수 있음
- FetchType이 Lazy로 설정된 다른 엔티티와 연관관계를 가지는 경우, 의도치 않은 `N+1` 문제가 발생할 수 있음
- FetchType이 Lazy로 설정된 다른 엔티티와 연관관계를 가지는 경우 트랜잭션 밖에서 접근 시 `LazyInitializationException` 발생할 수 있음
- 하위 계층인 `Persistence` 계층의 엔티티를 `Presentation` 계층의 DTO, Controller에 위치하는 것은 역할에 따라 계층을 분리하는 레이어드 아키텍쳐 관점에서 역방향으로 참조하고 있어 각 계층의 역할 경계를 흐트릴수 있고, 협업하는 다른 개발자로 하여금 혼선을 줄 여지가 있으므로 좋은 구조가 아님

re-mora committed:
refactor: PathFinder Class의 코드 정합성 및 가독성을 위한 제네릭 형식 명시

re-mora committed:
refactor: DefaultWeightedEdge 타입의 간선 표현을 SectionWeightedEdge 타입으로 변경

그래프의 간선을 어떤 객체로 표현하는지 보다 가독성있게 표현할 수 있도록 DefaultWeightedEdge를 상속 받은 신규 SectionWeightedEdge(구간간선) 타입 생성 및 변경

re-mora committed:
refactor: 하드코딩으로 명시된 오류 메세지 상수로 추출

re-mora committed:
refactor: 코드 정합성 및 가독성을 위한 제네릭 형식 명시

re-mora committed:
refactor: 최단 경로 응답 객체를 적절한 path package로 이동

re-mora committed:
refactor: static 변수로 할당된 공유 변수에 매번 값을 할당하면서 발생할 수 있는 쓰레드 세이프하지 않은 코드 수정

re-mora committed:
refactor: 자주 사용하는 Mokcing 대상 객체의 Fake 적용

LineRepository의 구현체로 FakeLineRepository를 재정의 및 Spy를 활용하여 자주 호출되는 테스트 조건에 대한 Mocking 조건 제거

re-mora commented:
안녕하세요 동휘님 :)

리뷰해주신 내용을 아래와 같이 정리하고 수정하였습니다!
의도하신 내용을 제가 적절하게 이해했는지 궁금합니다.

### 코드 리뷰 내용 정리
- 엔티티를 응답 객체로 반환하는 경우 발생할 수 있는 문제
  - 데이터베이스의 테이블과 엔티티를 반환하는 경우 테이블 명세가 외부로 노출
  - 응답 항목이 추가/변경 되는 요구사항이 있는 경우, 테이블과 매핑된 엔티티로는 유연하게 대처할 수 없음
  - 해당 엔티티가 양방향 연관관계를 가지고, toString이 구현되어 있는 경우 순환 참조로 인해 `StackOverFlow ` 발생할 수 있음에 주의
  - FetchType이 Lazy로 설정된 다른 엔티티와 연관관계를 가지는 경우, 의도치 않은 `N+1` 문제가 발생할 수 있음
  - FetchType이 Lazy로 설정된 다른 엔티티와 연관관계를 가지는 경우 트랜잭션 밖에서 접근 시 `LazyInitializationException` 발생할 수 있음에 주의
 
  - 하위 계층인 `Persistence` 계층의 엔티티를 `Presentation` 계층의 DTO, Controller에 위치하는 것은 역할에 따라 계층을 분리하는 레이어드 아키텍쳐 관점에서 역방향으로 참조하고 있어 각 계층의 역할 경계를 흐트릴수 있고, 협업하는 다른 개발자로 하여금 혼선을 줄 여지가 있으므로 좋은 구조가 아님

- 코드 정합성 및 가독성을 위한 제네릭 형식 명시

- static 변수로 할당된 공유 변수에 매번 값을 할당하면서 발생할 수 있는 쓰레드 세이프하지 않은 코드 수정

- Fake 객체를 활용하여 자주 정의되는 Mocking & Stubbing 코드 제거

미흡한 부분이 있다면 한번 더 피드백 부탁드리겠습니다!

---
리뷰해주신 내용이 생소한 부분도 있고, 이미 알고있었지만 습관적으로 지나치는 경우도 많아 반성도 하게되고 배우는 부분도 많은 것 같습니다.
꼼꼼하게 리뷰 해주셔서 너무 감사합니다. 

요 몇일 사이 비가 많이 오는데 항상 건강 조심하시고!
오늘도 좋은 하루 되세요 😄

dhmin5693 commented on src/main/java/nextstep/subway/path/dto/ShortestPathResponse.java:
```diff
@@ -0,0 +1,34 @@
+
+public class ShortestPathResponse {
+    private int distance;
+    private List<StationResponse> stations;
```
https://edu.nextstep.camp/s/X02BsEA0/ls/m8sHKr98
저번에 말씀드렸던 stations의 스펙이 맞춰지지 않았습니다.
강의 페이지를 참고하여 스펙을 맞춰주세요.
`StationResponse` 는 createdDate, modifiedDate가 있어서 다른 클래스를 만드는건 어떤가요?

---

## 강의에 기재된 스펙

```json
{
    "stations": [
        {
            "id": 5,
            "name": "양재시민의숲역",
            "createdAt": "2020-05-09T23:54:12.007"
        },
        {
            "id": 4,
            "name": "양재역",
            "createdAt": "2020-05-09T23:54:11.995"
        },
        {
            "id": 1,
            "name": "강남역",
            "createdAt": "2020-05-09T23:54:11.855"
        },
        {
            "id": 2,
            "name": "역삼역",
            "createdAt": "2020-05-09T23:54:11.876"
        },
        {
            "id": 3,
            "name": "선릉역",
            "createdAt": "2020-05-09T23:54:11.893"
        }
    ],
    "distance": 40
}
```

## 현재 스펙

```json
{
    "distance": 32,
    "stations": [
        {
            "id": 3,
            "name": "선정릉역",
            "createdDate": "2022-06-30T23:40:26.413482",
            "modifiedDate": "2022-06-30T23:40:26.413482"
        },
        {
            "id": 6,
            "name": "선릉역",
            "createdDate": "2022-06-30T23:40:26.483139",
            "modifiedDate": "2022-06-30T23:40:26.483139"
        },
        {
            "id": 5,
            "name": "역삼역",
            "createdDate": "2022-06-30T23:40:26.461796",
            "modifiedDate": "2022-06-30T23:40:26.461796"
        },
        {
            "id": 4,
            "name": "강남역",
            "createdDate": "2022-06-30T23:40:26.443782",
            "modifiedDate": "2022-06-30T23:40:26.443782"
        },
        {
            "id": 7,
            "name": "양재역",
            "createdDate": "2022-06-30T23:40:26.503396",
            "modifiedDate": "2022-06-30T23:40:26.503396"
        }
    ]
}
```

dhmin5693 commented on src/main/java/nextstep/subway/path/domain/PathFinder.java:
```diff
@@ -0,0 +1,93 @@
+
+    private void addVertexByStation(WeightedMultigraph<Station, SectionWeightedEdge> graph, List<Station> allLinesStations) {
+        allLinesStations.forEach(graph::addVertex);
+    }
```
graph를 필드로 바꾸고 생성자에서 추가해주면 불변을 지켜줄 수 있으며 여러 번 사용할 때도 성능 상 이점이 있을 것 같습니다.

그리고 지금은 graph와 list를 매개변수로 받아서 addVertex로 추가를 해주고 있어요.
graph가 컬렉션은 아니지만 컬렉션처럼 내부에 데이터를 저장하고 활용한다는 점은 같습니다.
컬렉션을 매개변수로 받고 그 안에 데이터를 삽입하거나 삭제하는 방식의 코드는 유지보수할 때 디버깅을 어렵게 만드는 방법이라 지양하셨으면 좋겠어요.

극단적인 예시를 들어볼게요.
Map을 서비스에서 생성합니다. 이 Map은 동적 쿼리로 사용하기 위해 dao까지 넘겨야 합니다.
상당히 오래된 레거시 코드라서 서비스 -> 다른 서비스 -> 다른 서비스 -> dao -> dao 식으로 5번 이동하는 동안 각 클래스의 private 메소드도 거친다고 해볼게요.
수많은 메소드를 거치다보면 특정 데이터가 어느 순간에 `Map`에 인입되었는지 확인하기 어렵지 않을까요?

dhmin5693 reviewed (changes requested):
피드백 반영을 잘 해주셨는데요, 가장 중요한 "강의 상의 요구사항" 부분이 빠져있어요.
다시 한번 확인 부탁드릴게요.
어짜피 요청드리는 겸, 생각난 점 몇 가지를 더 써두었으니 같이 확인해주세요! 🙏

re-mora committed:
fix: API 요구사항에 맞도록 신규 응답 객체 생성

re-mora committed:
refactor: graph 객체를 필드로 변경하고 생성자에서 값을 할당하도록 변경

re-mora committed:
refactor: 최단 경로 조회 객체 생성 시점에 탐색 경로를 설정하도록 변경

re-mora committed:
refactor: 리팩토링 이후 더 이상 참조가 없는 메서드 제거

re-mora commented:
안녕하세요, 동휘님 :)
리뷰해주신 내용을 아래와 같이 수정해보았습니다.

### API 응답 객체 생성 
이 부분은 드릴 말씀이 없습니다.. ㅠ 가장 중요한 요구사항이었음에도 불구하고, 로직 구현에만 신경쓰다보니 미처 확인하지 못한것 같습니다. 신규 객체인 `StationSimpleResponse` 객체를 생성하여 요구사항 스펙에 맞게 반환하도록 수정하였습니다.

```json
{
    "stations": [
        {
            "id": 3,
            "name": "선정릉역",
            "createdAt": "2022-07-01T00:53:03.962"
        },
        {
            "id": 6,
            "name": "선릉역",
            "createdAt": "2022-07-01T00:53:04.024"
        },
        {
            "id": 5,
            "name": "역삼역",
            "createdAt": "2022-07-01T00:53:04.009"
        },
        {
            "id": 4,
            "name": "강남역",
            "createdAt": "2022-07-01T00:53:03.993"
        },
        {
            "id": 7,
            "name": "양재역",
            "createdAt": "2022-07-01T00:53:04.055"
        }
    ],
    "distance": 32
}
```

### 최단 경로 조회 객체 `PathFinder` 의 구성 변경
이 부분은 말씀하신 컬렉션 객체가 로직 구현과정에서 이동 될 수 있고, 그 과정 에서 여러번 가변되어 추적하기 어려운 현상을 말씀하신것 같아, 아래와 같이 수정하였습니다.

##### 최단 경로 객체 생성 시, 경로를 탐색할 구간을 주입받도록 수정
- 주입 받은 탐색 구간을 이용하여 최단 경로 객체 생성 시점에 그래프에 정점과 간선을 설정
- 외부에서 출발지와 목적지만 입력받아 값 설정이 완료된 그래프 내에서 최단 경로를 탐색
- 동일 탐색 구간에서 출발지와 목적지만 입력받아 추가 객체 생성 없이 여러번 탐색가능

```java
    public ShortestPathResponse getShortestPath(Long sourceStationId, Long targetStationId) {
        ...
        PathFinder pathFinder = PathFinder.from(allLines);
        return pathFinder.findShortestPath(sourceStation, targetStation);
    }
```

```java
    private PathFinder(List<Line> allLines) {
        this.graph = new WeightedMultigraph<>(SectionWeightedEdge.class);
        this.dijkstraShortestPath = new DijkstraShortestPath<>(graph);
        this.allLines = allLines;
        addSectionsToGraph();
    }

    public static PathFinder from(List<Line> allLines) {
        return new PathFinder(allLines);
    }
```

- 리팩터링 이전 : 컬렉션을 파라미터로 받아 내부에서 가공 할 수 있으므로 말씀하신 것처럼 어디서 언제 변경되는지 추적이 어려움
- 리팩터링 이후 : 객체 생성 시점에 주입받은 컬렉션을 필드로 활용하여, 컬렉션을 파라미터로 하는 메서드 제거

부족한 점이 많지만, 이번 수정 사항 역시 미흡한 부분이 있다면 한번 더 피드백 부탁드리겠습니다. 

---

늦은 시간까지 꼼꼼하게 봐주셔서 다시 한번 감사합니다 :)

dhmin5693 commented on src/main/java/nextstep/subway/path/domain/PathFinder.java:
```diff
@@ -54,16 +58,16 @@ private void validateStationContains(List<Line> allLines, Station source, Statio
+    private void addSectionsToGraph(List<Line> allLines) {
+        addVertexByStation(mergeAllLinesStations(allLines));
+        setEdgeWeightBySection( mergeAllLinesSections(allLines));
     }
```
저는 이 메소드까지 생성자에 넣는게 좋다고 생각해서 드린 리뷰였어요!
약간 오해의 소지가 있게끔 남겼었네요 😅
생성할 때 `Line` 정보도 밀어넣고 생성된 뒤에는 건드리지 않아야 불변성을 지킬 수 있거든요.

이대로도 상관없지만 용식님께서 생각하시기에 바꾸는게 더 낫다 싶으면 다음 미션에 반영해주세요 :)

dhmin5693 reviewed (approved):
피드백 잘 반영해주셨네요 :)
다음 미션 진행해주세요.
다음 미션도 화이팅입니다.
